hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java

```diff
@@ -214,7 +261,7 @@ protected AbstractSerializationService(AbstractSerializationService prototype) {
             throw handleException(e);
         } finally {
             ClassLocator.onFinishDeserialization();
-            pool.returnInputBuffer(in);
+            serializer.conditionallyReturnInputBufferToPool(obj, in, pool);
```

Comment:
It would lead to more stress on the GC. I can not see any other problem.
 The alternative is to copy the byte-array for the cases where we return GenericRecord (implementation is DefaultCompactReader). But that would lead to GC as well together with extra copy cost.